server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -129,7 +130,13 @@ public void close() {
      */
     public StoredContext stashContext() {
         final ThreadContextStruct context = threadLocal.get();
-        threadLocal.set(null);
+        if(context.requestHeaders.containsKey(Task.X_OPAQUE_ID)) {
```

Comment:
If we must have this special case, then there should at least be a good explanation in comment on why it is necessary.

minor nit: there are spacing issues in the if and else lines

Reply:
added comment and fixed spacing